File: mqtt-mapping-service/src/main/java/mqtt/mapping/configuration/ConnectorConfigurationComponent.java
/*
 * Copyright (c) 2022 Software AG, Darmstadt, Germany and/or Software AG USA Inc., Reston, VA, USA,
 * and/or its subsidiaries and/or its affiliates and/or their licensors.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @authors Christof Strack, Stefan Witschel
 */

package mqtt.mapping.configuration;

import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.model.option.OptionPK;
import com.cumulocity.rest.representation.tenant.OptionRepresentation;
import com.cumulocity.sdk.client.SDKException;
import com.cumulocity.sdk.client.option.TenantOptionApi;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Slf4j
@Component
public class ConnectorConfigurationComponent {
    private static final String OPTION_CATEGORY_CONFIGURATION = "dynamic.mapper.service";
    private static final String OPTION_KEY_CONNECTION_CONFIGURATION = "credentials.connection.configuration";
    private static final String OPTION_KEY_SERVICE_CONFIGURATION = "service.configuration";

    private final TenantOptionApi tenantOptionApi;

    @Autowired
    private MicroserviceSubscriptionsService subscriptionsService;

    private ObjectMapper objectMapper;

    @Autowired
    public void setObjectMapper(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @Autowired
    public ConnectorConfigurationComponent(TenantOptionApi tenantOptionApi) {
        this.tenantOptionApi = tenantOptionApi;
    }

    public String getConnectorOptionKey(String ident) {
        return OPTION_KEY_CONNECTION_CONFIGURATION + "." + ident;
    }

    public void saveConnectorConfiguration(final ConnectorConfiguration configuration)
            throws JsonProcessingException {
        if (configuration == null) {
            return;
        }
        String ident = configuration.getIdent();
        final String configurationJson = objectMapper.writeValueAsString(configuration);
        final OptionRepresentation optionRepresentation = OptionRepresentation
                .asOptionRepresentation(OPTION_CATEGORY_CONFIGURATION, getConnectorOptionKey(ident), configurationJson);
        tenantOptionApi.save(optionRepresentation);
    }

    public void deleteConnectionConfiguration(final String ident)
            throws JsonProcessingException {
        if (ident == null) {
            return;
        }
        final OptionPK option = new OptionPK();
        option.setCategory(OPTION_CATEGORY_CONFIGURATION);
        option.setKey(getConnectorOptionKey(ident));
        tenantOptionApi.delete(option);
    }

    public ConnectorConfiguration getConnectorConfiguration(String ident, String tenant) {
        final OptionPK option = new OptionPK();
        option.setCategory(OPTION_CATEGORY_CONFIGURATION);
        option.setKey(getConnectorOptionKey(ident));
        ConnectorConfiguration result = subscriptionsService.callForTenant(tenant, () -> {
            ConnectorConfiguration rt = null;
            try {
                final OptionRepresentation optionRepresentation = tenantOptionApi.getOption(option);
                final ConnectorConfiguration configuration = new ObjectMapper().readValue(
                        optionRepresentation.getValue(),
                        ConnectorConfiguration.class);
                log.debug("Returning connection configuration found: {}:", configuration.getConnectorId());
                rt = configuration;
            } catch (SDKException exception) {
                log.warn("Tenant {} - No configuration found, returning empty element!", tenant);
                rt = null;
            } catch (JsonMappingException e) {
                e.printStackTrace();
            } catch (JsonProcessingException e) {
                e.printStackTrace();
            }
            return rt;
        });
        return result;
    }

    public List<ConnectorConfiguration> getConnectorConfigurations(String tenant) {
        final OptionPK option = new OptionPK();
        final List<ConnectorConfiguration> connectorConfigurations = new ArrayList<>();
        subscriptionsService.runForTenant(tenant, () -> {
            try {
                final List<OptionRepresentation> optionRepresentationList = tenantOptionApi
                        .getAllOptionsForCategory(OPTION_CATEGORY_CONFIGURATION);
                for (OptionRepresentation optionRepresentation : optionRepresentationList) {
                    // Just Connector Config --> Ignoring Service Configuration
                    String optionKey = OPTION_KEY_CONNECTION_CONFIGURATION.replace("credentials.", "");
                    if (optionRepresentation.getKey().startsWith(optionKey)) {
                        final ConnectorConfiguration configuration = new ObjectMapper().readValue(
                                optionRepresentation.getValue(),
                                ConnectorConfiguration.class);
                        connectorConfigurations.add(configuration);
                        log.debug("Connection configuration found: {}:", configuration.getConnectorId());
                    }
                }
            } catch (SDKException exception) {
                log.warn("Tenant {} - No configuration found, returning empty element!", tenant);
            } catch (JsonMappingException e) {
                e.printStackTrace();
            } catch (JsonProcessingException e) {
                e.printStackTrace();
            }
        });
        return connectorConfigurations;
    }

    public void deleteConcetorConfigurations(String tenant) {
        List<ConnectorConfiguration> configs = getConnectorConfigurations(tenant);
        for (ConnectorConfiguration config : configs) {
            OptionPK optionPK = new OptionPK(OPTION_CATEGORY_CONFIGURATION,
                    getConnectorOptionKey(config.getConnectorId()));
            tenantOptionApi.delete(optionPK);
        }
        OptionPK optionPK = new OptionPK(OPTION_CATEGORY_CONFIGURATION, OPTION_KEY_SERVICE_CONFIGURATION);
        tenantOptionApi.delete(optionPK);
    }

    public ConnectorConfiguration enableConnection(String connectorId, boolean enabled) {
        final OptionPK option = new OptionPK(OPTION_CATEGORY_CONFIGURATION, getConnectorOptionKey(connectorId));
        try {
            final OptionRepresentation optionRepresentation = tenantOptionApi.getOption(option);
            final ConnectorConfiguration configuration = new ObjectMapper().readValue(optionRepresentation.getValue(),
                    ConnectorConfiguration.class);
            configuration.enabled = enabled;
            log.debug("Setting connection: {}:", configuration.enabled);
            final String configurationJson = new ObjectMapper().writeValueAsString(configuration);
            optionRepresentation.setCategory(OPTION_CATEGORY_CONFIGURATION);
            optionRepresentation.setKey(getConnectorOptionKey(connectorId));
            optionRepresentation.setValue(configurationJson);
            tenantOptionApi.save(optionRepresentation);
            return configuration;
        } catch (SDKException exception) {
            log.warn("No configuration found, returning empty element!");
            // exception.printStackTrace();
        } catch (JsonMappingException e) {
            e.printStackTrace();
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
        return null;
    }

}

File: dynamic-mapping-ui/src/shared/connector-configuration/connector-configuration.component.ts
/*
 * Copyright (c) 2022 Software AG, Darmstadt, Germany and/or Software AG USA Inc., Reston, VA, USA,
 * and/or its subsidiaries and/or its affiliates and/or their licensors.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @authors Christof Strack
 */
import { Component, OnInit } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { Observable } from 'rxjs';
import packageJson from '../../../package.json';

import * as _ from 'lodash';
import { ConfirmationModalComponent } from '../confirmation/confirmation-modal.component';
import { ConnectorConfigurationService } from '../connector-configuration.service';
import {
  ConnectorStatus,
  StatusEventTypes
} from '../connector-status/connector-status.model';
import { Direction, Feature } from '../model/shared.model';
import { uuidCustom } from '../model/util';
import { Operation } from '../shared.module';
import { SharedService } from '../shared.service';
import { ConfigurationConfigurationModalComponent } from './connector-configuration-modal.component';
import {
  ConnectorConfiguration,
  ConnectorSpecification
} from './connector.model';

@Component({
  selector: 'd11r-mapping-connector-configuration',
  styleUrls: ['./connector-configuration.component.style.css'],
  templateUrl: 'connector-configuration.component.html'
})
export class ConnectorConfigurationComponent implements OnInit {
  version: string = packageJson.version;
  monitoring$: Observable<ConnectorStatus>;
  feature: Feature;
  specifications: ConnectorSpecification[] = [];
  configurations: ConnectorConfiguration[];
  StatusEventTypes = StatusEventTypes;

  constructor(
    public bsModalService: BsModalService,
    public connectorConfigurationService: ConnectorConfigurationService,
    public alertService: AlertService,
    private sharedService: SharedService
  ) {}

  async ngOnInit() {
    // console.log('Running version', this.version);
    this.feature = await this.sharedService.getFeatures();
    if (!this.feature.userHasMappingAdminRole) {
      this.alertService.warning(
        "The configuration on this tab is not editable, as you don't have Mapping ADMIN permissions. Please assign Mapping ADMIN permissions to your user."
      );
    }
    this.specifications =
      await this.connectorConfigurationService.getConnectorSpecifications();
    this.connectorConfigurationService
      .getConnectorConfigurationsLive()
      .subscribe((confs) => {
        this.configurations = confs;
      });

    await this.loadData();
  }

  async refresh() {
    this.connectorConfigurationService.resetCache();
    await this.loadData();
  }
  async loadData(): Promise<void> {
    await this.connectorConfigurationService.startConnectorConfigurations();
  }

  async clickedReconnect2NotificationEndpoint() {
    const response1 = await this.sharedService.runOperation(
      Operation.REFRESH_NOTIFICATIONS_SUBSCRIPTIONS
    );
    // console.log('Details reconnect2NotificationEndpoint', response1);
    if (response1.status === 201) {
      this.alertService.success(gettext('Reconnected successfully.'));
    } else {
      this.alertService.danger(gettext('Failed to reconnect!'));
    }
  }

  async onConfigurationUpdate(index) {
    const configuration = this.configurations[index];

    const initialState = {
      add: false,
      configuration: configuration,
      specifications: this.specifications
    };
    const modalRef = this.bsModalService.show(
      ConfigurationConfigurationModalComponent,
      {
        initialState
      }
    );
    modalRef.content.closeSubject.subscribe(async (editedConfiguration) => {
      // console.log('Configuration after edit:', editedConfiguration);
      if (editedConfiguration) {
        this.configurations[index] = editedConfiguration;
        // avoid to include status$
        const clonedConfiguration = {
          ident: editedConfiguration.ident,
          connectorType: editedConfiguration.connectorType,
          enabled: editedConfiguration.enabled,
          name: editedConfiguration.name,
          properties: editedConfiguration.properties
        };
        const response =
          await this.connectorConfigurationService.updateConnectorConfiguration(
            clonedConfiguration
          );
        if (response.status < 300) {
          this.alertService.success(gettext('Updated successfully.'));
        } else {
          this.alertService.danger(
            gettext('Failed to update connector configuration')
          );
        }
      }
      await this.loadData();
    });
  }

  async onConfigurationCopy(index) {
    const configuration = _.clone(this.configurations[index]);
    configuration.ident = uuidCustom();
    configuration.name = `${configuration.name}_copy`;
    this.alertService.warning(
      gettext(
        'Review properties, e.g. client_id must be different across different client connectors to the same broker.'
      )
    );

    const initialState = {
      add: false,
      configuration: configuration,
      specifications: this.specifications
    };
    const modalRef = this.bsModalService.show(
      ConfigurationConfigurationModalComponent,
      {
        initialState
      }
    );
    modalRef.content.closeSubject.subscribe(async (editedConfiguration) => {
      // console.log('Configuration after edit:', editedConfiguration);
      if (editedConfiguration) {
        this.configurations[index] = editedConfiguration;
        // avoid to include status$
        const clonedConfiguration = {
          ident: editedConfiguration.ident,
          connectorType: editedConfiguration.connectorType,
          enabled: editedConfiguration.enabled,
          name: editedConfiguration.name,
          properties: editedConfiguration.properties
        };
        const response =
          await this.connectorConfigurationService.createConnectorConfiguration(
            clonedConfiguration
          );
        if (response.status < 300) {
          this.alertService.success(gettext('Updated successfully.'));
        } else {
          this.alertService.danger(
            gettext('Failed to update connector configuration!')
          );
        }
      }
      await this.loadData();
    });
  }

  async onConfigurationDelete(index) {
    const configuration = this.configurations[index];

    const initialState = {
      title: 'Delete connector',
      message: 'You are about to delete a connector. Do you want to proceed?',
      labels: {
        ok: 'Delete',
        cancel: 'Cancel'
      }
    };
    const confirmDeletionModalRef: BsModalRef = this.bsModalService.show(
      ConfirmationModalComponent,
      { initialState }
    );
    confirmDeletionModalRef.content.closeSubject.subscribe(
      async (result: boolean) => {
        // console.log('Confirmation result:', result);
        if (result) {
          const response =
            await this.connectorConfigurationService.deleteConnectorConfiguration(
              configuration.ident
            );
          if (response.status < 300) {
            this.alertService.success(gettext('Deleted successfully.'));
          } else {
            this.alertService.danger(
              gettext('Failed to delete connector configuration')
            );
          }
          await this.loadData();
        }
        confirmDeletionModalRef.hide();
      }
    );
    await this.loadData();
  }

  async onConfigurationToggle(index) {
    const configuration = this.configurations[index];
    const response1 = await this.sharedService.runOperation(
      configuration.enabled ? Operation.DISCONNECT : Operation.CONNECT,
      { connectorIdent: configuration.ident }
    );
    // console.log('Details toggle activation to broker', response1);
    if (response1.status === 201) {
      // if (response1.status === 201 && response2.status === 201) {
      this.alertService.success(gettext('Connection updated successfully.'));
    } else {
      this.alertService.danger(gettext('Failed to establish connection!'));
    }
    await this.loadData();
    this.sharedService.refreshMappings(Direction.INBOUND);
    this.sharedService.refreshMappings(Direction.OUTBOUND);
  }
}
